jellyfin-core/src/jvmCommonMain/kotlin/org/jellyfin/sdk/discovery/LocalServerDiscovery.kt

```diff
@@ -0,0 +1,131 @@
+package org.jellyfin.sdk.discovery
+
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.flow
+import kotlinx.coroutines.isActive
+import kotlinx.serialization.SerializationException
+import kotlinx.serialization.json.Json
+import mu.KotlinLogging
+import org.jellyfin.sdk.JellyfinOptions
+import org.jellyfin.sdk.model.api.ServerDiscoveryInfo
+import java.io.IOException
+import java.net.DatagramPacket
+import java.net.DatagramSocket
+import java.net.InetAddress
+import java.net.SocketTimeoutException
+import kotlin.coroutines.coroutineContext
+
+private val logger = KotlinLogging.logger {}
+
+/**
+ * Used to discover Jellyfin servers in the local network.
+ *
+ * Use the [discover] function to retrieve a flow of servers until the timeout is exceeded or
+ * the maximum amount of servers has been retrieved.
+ */
+public actual class LocalServerDiscovery actual constructor(jellyfinOptions: JellyfinOptions) {
+	private val discoveryBroadcastAddressesProvider = DiscoveryBroadcastAddressesProvider(jellyfinOptions)
+
+	public actual companion object {
+		public const val DISCOVERY_MESSAGE: String = "who is JellyfinServer?"
+		public const val DISCOVERY_PORT: Int = 7359
+		public const val DISCOVERY_RECEIVE_BUFFER: Int = 1024 // bytes
+		public actual val DISCOVERY_TIMEOUT: Int = 500 // ms
+		public actual val DISCOVERY_MAX_SERVERS: Int = 15
+	}
+
+	private val json = Json {
+		ignoreUnknownKeys = true
+	}
+
+	/**
+	 * Send our broadcast message to a given address
+	 */
+	private fun discoverAddress(socket: DatagramSocket, address: InetAddress): Unit = try {
+		val message = DISCOVERY_MESSAGE.toByteArray()
+		val packet = DatagramPacket(message, message.size, address, DISCOVERY_PORT)
+		socket.send(packet)
+
+		logger.debug { "Discovering via $address" }
+	} catch (err: IOException) {
+		logger.error(err) { "Unable to send discovery message to $address" }
```

Comment:
Don't all these `{ "..." }` logging calls have more overhead since they need to execute another function? I'd only use them if we really need the lazy evaluation honestly.

Reply:
They are only evaluated if the logging level is enabled (which it is most of the time) but there are no functions to just pass a string in kotlin-logging common.

Reply:
Weird decision for their API. Not a fan of using lazy evaluation everywhere even if we have logging enabled generally, but it looks like I have to accept that.